ksql-version-metrics-client/src/test/java/io/confluent/ksql/version/metrics/KsqlVersionCheckerAgentTest.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.version.metrics;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.verify;
+
+import io.confluent.ksql.version.metrics.collector.KsqlModuleType;
+import io.confluent.support.metrics.BaseSupportConfig;
+import java.time.Clock;
+import java.util.Properties;
+import java.util.function.Supplier;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class KsqlVersionCheckerAgentTest {
+
+  @Mock
+  private Clock clock;
+  @Mock
+  private KsqlVersionChecker ksqlVersionChecker;
+  @Mock
+  private Properties properties;
+
+  private KsqlVersionCheckerAgent ksqlVersionCheckerAgent;
+
+  @Before
+  public void setup() {
+    ksqlVersionCheckerAgent = new KsqlVersionCheckerAgent(
+        () -> true,
+        true,
+        clock,
+        (BaseSupportConfig ksqlVersionCheckerConfig,
+            KsqlModuleType moduleType,
+            boolean enableSettlingTime,
+            Supplier<Boolean> activenessStatusSupplier) -> ksqlVersionChecker);
+  }
+
+  @Test
+  public void shouldStartTheAgentCorrectly() throws Exception {
+    // When:
+    ksqlVersionCheckerAgent.start(KsqlModuleType.SERVER, properties);
+
+    // Then:
+    verify(ksqlVersionChecker).init();
+    verify(ksqlVersionChecker).setUncaughtExceptionHandler(any());
+    verify(ksqlVersionChecker).start();
+  }
+
+  @Test (expected = Exception.class)
+  public void shouldSFailIfVersionCheckerFails() throws Exception {
+    // Given:
+    doThrow(new Exception("FOO")).when(ksqlVersionChecker).start();
+
+    // When:
+    ksqlVersionCheckerAgent.start(KsqlModuleType.SERVER, properties);
+  }
+
+
+  @Test
+  public void shouldUpdateLastRequestTimeCorrectly() throws Exception {
```

Comment:
This test is pretty pointless. It proves the code calls `clock.millis()`, but it could do anything or nothing with the result.

What you need in here is a few tests to test:
- if there are no active queries and no recent requests, then the isActive supplier passed to the versionCheckerFactory returns false.
- if there was a request, but it was ages ago, it still returns false.
- if there are active queries it returns true.
- if there was a recent request it returns true.

To do this, you will need to be able to _capture_ the isActive supplier passed to the version checker factory.  We can pair on this if you would like.

Reply:
Tests were added.